### PR TITLE
Fix readNullable on root path (issue #11)

### PR DESCRIPTION
**Pull request by pbugnion**
This PR is an attempt at addressing issue #11 .

Prior to this PR, the calling `__.readNullable[T]` would return 'error.path.missing' even when the root object was correctly formed. This commit modifies the code so that it works, at least when the object is present.

With this PR, the following works:

```scala
import play.api.libs.functional.syntax._
import play.api.libs.json._

case class User(name: String, age: Int)
case class Address(street: String, postcode: String)

implicit val readsUser: Reads[User] = (
  (__ \ "name").read[String] and (__ \ "age").read[Int]
)(User)
implicit val readsAddress: Reads[Address] = (
  (__ \ "street").read[String] and (__ \ "postcode").read[String]
)(Address)

implicit val readsUserAddress: Reads[(User, Option[Address])] = (
  __.read[User] and __.readNullable[Address]
).tupled

val sherlockHolmes = Json.obj(
  "name" -> "Sherlock Holmes",
  "age" -> 32,
  "street" -> "221B Baker Street",
  "postcode" -> "NW1 6XE"
)
sherlockHolmes.validate(readsUserAddress)
// (User("Sherlock Holmes", 32), Some(Address("221B Baker Street", "NW1 6XE")))
```

While this works, I'm not sure how to address the case where specific fields are missing. For instance, if `sherlockHolmes` just contains `User` fields:

```
val sherlockHolmes = Json.obj(
  "name" -> "Sherlock Holmes",
  "age" -> 32
)
sherlockHolmes.validate(readsUserAddress)
// JsError(List((/postcode,List(JsonValidationError(List(error.path.missing),WrappedArray()))), (/street,List(JsonValidationError(List(error.path.missing),WrappedArray())))))
```

Is this the behaviour we expect, or should this be `JsSuccess(User(...), None)`?

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @pbugnion,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

File: play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala [hunk 1]
```diff
@@ -216,6 +216,49 @@ class JsonValidSpec extends Specification {
       JsString("alphabeta").validate[String] must equalTo(JsSuccess("alphabeta"))
     }
 
+    "validate reads on the root path" in {
```
Comment (cchantep) [inline play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala]: `".." >> { .. }` (No `in` in `in`)

File: play-json/src/main/scala/play/api/libs/json/JsPath.scala [hunk 2]
```diff
@@ -195,7 +195,7 @@ case class JsPath(path: List[PathNode] = List()) {
 
   def applyTillLast(json: JsValue): Either[JsError, JsResult[JsValue]] = {
     def step(path: List[PathNode], json: JsValue): Either[JsError, JsResult[JsValue]] = path match {
```
Comment (cchantep) [inline play-json/src/main/scala/play/api/libs/json/JsPath.scala]: `@annotation.tailrec` could also be added

File: play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala (hunk 1 above)
Comment (pbugnion) [inline play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala]: As in, I should write:

```scala
"validate reads on the root path" in {
  "reads" >> { ... }
  "readNullables" >> { ... }
}
```

or 

```scala
"validate reads on the root path" >> {
  "reads" in { ... }
  "readNullables" in { ... }
}
```
?

File: play-json/src/main/scala/play/api/libs/json/JsPath.scala (hunk 2 above)
Comment (pbugnion) [inline play-json/src/main/scala/play/api/libs/json/JsPath.scala]: Done

File: play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala (hunk 1 above)
Comment (cchantep) [inline play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala]: Second one

Comment (pbugnion) [inline play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala]: Done -- I've also changed some other tests in the same file that applied the same pattern.

File: play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala [hunk 3]
```diff
@@ -216,6 +216,49 @@ class JsonValidSpec extends Specification {
...
+        )
+      }
+
+      "readNullables" in {
+        implicit val userAddressReads: Reads[(User, Option[Address])] = (
+          __.read[User] and
+          __.readNullable[Address]
+        ).tupled
+
+        bobby.validate[(User, Option[Address])] must equalTo(
```
Comment (cchantep) [inline play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala]: Needs another test without `street` and `zip` in the JSON object, with expected `JsSuccess` with `None` address

Comment (pbugnion) [inline play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala]: I've replied as a comment below.

**Comment by pbugnion** [comment]
It's not obvious (at least to me) what the correct behaviour should be when the root path is an object, but contains the wrong field. For instance, with the following reads:

```scala
case class User(name: String, age: Int)
case class Address(street: String, postcode: String)

implicit val readsUser: Reads[User] = (
  (__ \ "name").read[String] and (__ \ "age").read[Int]
)(User)
implicit val readsAddress: Reads[Address] = (
  (__ \ "street").read[String] and (__ \ "postcode").read[String]
)(Address)

implicit val readsUserAddress: Reads[(User, Option[Address])] = (
  __.read[User] and __.readNullable[Address]
).tupled
```

If the body is missing the address fields:

```scala
val sherlockHolmes = Json.obj(
  "name" -> "Sherlock Holmes",
  "age" -> 32
)
```

then: `sherlockHolmes.validate(readsUserAddress)` could be expected to return either `JsSuccess(None)` or `JsError()` corresponding to a missing address. While returning `JsSuccess(None)` seems more intuitive, it breaks some invariants. Currently (and prior to this PR), if we have:

```scala
val nestedSherlockHolmes = Json.obj("user" -> sherlockHolmes)

(nestedSherlockHolmes \ "user").validate(readsUserAddress)
// Returns JsError because the address cannot be read
```

The only time that you get `JsSuccess(None)` is if the path is missing, or the root is `JsNull`:

```scala
implicit val readsUserAddressOpts: Reads[Option[(User, Address)]] = (
  __.readNullable[User] and __.readNullable[Address]
).tupled

JsNull.validate(readsUserAddressOpts) // prior to this PR, would return JsError
// JsSuccess((None, None))

(Json.obj("user" -> JsNull) \ "user").validate(readsUserAddressOpts)
// JsSuccess((None, None))
```

**Comment by cchantep** [comment]
For me, it's coherent to raise a`JsError` when reading an invalid (because empty or not containing the required field) object, wherever it is, at root or itself nested.

Your opinion @gmethvin ?

**Comment by gmethvin** [comment]
There are two cases:
 1. The value to read is null or missing (e.g. the field is missing), so we return `JsSuccess(None)`
 2. The value exists, so we validate it and return the result of that validation as `JsSuccess(Some(result))` if successful, or the `JsError` if it failed.

So if we apply a `readNullable` to an invalid value (that is non-null or non-empty), it should return `JsError`.

**Comment by cchantep** [comment]
So we agree the test to be added there must check the appropriate `JsError` is returned.

**Comment by gmethvin** [comment]
correct

**Comment by pbugnion** [comment]
Thanks a lot for the feedback. I'll add a test.

**Comment by pbugnion** [comment]
I've added two unit tests, one to validate reading from a null root path and one from a badly formed root path.

File: play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala [hunk 4]
```diff
@@ -216,6 +216,77 @@ class JsonValidSpec extends Specification {
...
+          "age" -> 54,
+          "street" -> "13 Main St"
+        )
+
+        missingZipBobby.validate(userAddressReads) must equalTo(
+          JsError(__ \ "zip", JsonValidationError("error.path.missing"))
+        )
+      }
+
+      "readNullables for null root path" >> {
```
Comment (cchantep) [inline play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala]: This test with `JsNull` can be kept, but I had in mind a test with a non-null/non-empty root but with no field for address (to ensure it's considered in the same way as an empty object.

Comment (pbugnion) [inline play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala]: The commit adds *two* tests. Besides the `JsNull` one, it adds:

```scala
"readNullables for badly formed root path" >> {
  implicit val userAddressReads: Reads[(User, Option[Address])] = (
    __.read[User] and
    __.readNullable[Address]
  ).tupled

  val missingZipBobby = Json.obj(
    "name" -> "bobby",
    "age" -> 54,
    "street" -> "13 Main St"
  )]
}
```

Comment (cchantep) [inline play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala]: I means a test with `Json.obj("name" -> "bobby", "age" -> 54)` ("no field for address"), which corresponds to test validating with an empty object (excluding fields for user) rather than validating an invalid object (with street but missing zip).

Comment (pbugnion) [inline play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala]: Ah that makes sense. I'll add that test.

Comment (pbugnion) [inline play-json/src/test/scala/play/api/libs/json/JsonValidSpec.scala]: I've added a unit test.
